src/encodings/levels.rs

```diff
@@ -15,47 +15,80 @@
 use util::memory::ByteBufferPtr;
 use super::rle_encoding::{RleEncoder, RleDecoder};
 
-
-/// A encoder for definition/repetition levels. This is a thin
-/// wrapper on `RleEncoder`. Currently only support RLE encoding.
-pub struct LevelEncoder {
-  encoding: Encoding,
-  bit_width: u8,
-  rle_encoder: RleEncoder
+/// A encoder for definition/repetition levels.
+/// Currently only supports RLE and BIT_PACKED (dev/null) encoding.
+pub enum LevelEncoder {
+  RLE { bit_width: u8, rle_encoder: RleEncoder },
+  BIT_PACKED { bit_width: u8, bit_packed_encoder: BitWriter },
 }
 
 impl LevelEncoder {
+  /// Create new level encoder based on encoding, max level and underlying byte buffer.
+  /// For bit packed encoding it is assumed that buffer is already allocated with
+  /// 'LevelEncoder::max_buffer_size' method.
+  ///
+  /// Panics, if encoding is not supported
   pub fn new(encoding: Encoding, max_level: i16, byte_buffer: Vec<u8>) -> Self {
-    assert!(encoding == Encoding::RLE, "Currently only support RLE encoding");
     let bit_width = log2(max_level as u64 + 1) as u8;
-
-    Self {
-      encoding: encoding,
-      bit_width: bit_width,
-      rle_encoder: RleEncoder::new_from_buf(bit_width, byte_buffer, mem::size_of::<i32>())
+    match encoding {
+      Encoding::RLE => {
+        LevelEncoder::RLE {
+          bit_width: bit_width,
+          rle_encoder: RleEncoder::new_from_buf(bit_width, byte_buffer, mem::size_of::<i32>())
+        }
+      },
+      Encoding::BIT_PACKED => {
+        // here we set full byte buffer without adjusting for num_buffered_values, because
+        // byte buffer will already be allocated with size from 'max_buffer_size()' method.
+        LevelEncoder::BIT_PACKED {
+          bit_width: bit_width,
+          bit_packed_encoder: BitWriter::new_from_buf(byte_buffer, 0)
+        }
+      },
+      _ => panic!("Unsupported encoding type {}", encoding)
     }
   }
 
+  /// Put/encode levels vector into this level encoder.
+  /// Returns number of encoded values that are less than or equal to length of the input buffer.
+  ///
+  /// RLE and BIT_PACKED level encoders return Err() when internal buffer overflows or flush fails.
   #[inline]
-  pub fn put(&mut self, buffer: &[i16]) -> Result<i16> {
+  pub fn put(&mut self, buffer: &[i16]) -> Result<usize> {
     let mut num_encoded = 0;
-    for v in buffer {
-      if !self.rle_encoder.put(*v as u64)? { break; }
-      num_encoded += 1;
+    match *self {
+      LevelEncoder::RLE { bit_width: _, ref mut rle_encoder } => {
+        for v in buffer {
+          if !rle_encoder.put(*v as u64)? { break; }
+          num_encoded += 1;
+        }
+        rle_encoder.flush()?;
+      },
+      LevelEncoder::BIT_PACKED { bit_width, ref mut bit_packed_encoder } => {
+        for v in buffer {
+          if !bit_packed_encoder.put_value(*v as u64, bit_width as usize) {
+            return Err(general_err!("Not enough bytes left"));
```

Comment:
This is similar to RLE, but in RLE the error is thrown from `rle_encoding.rs`.

Reply:
Makes sense. I think `rle_encoder` returns `Err` when the internal buffer is full, which is confusing. I was not consistent in the return value: when the buffer is full, sometimes we return false, while sometimes we return `Err`. In future maybe we should add a new error type such as `BufferOverflow`. But this doesn't have to be addressed in this PR.